similarity-data-rest/src/main/java/org/phenotips/data/similarity/rest/internal/DefaultPatientMatchResource.java

```diff
@@ -0,0 +1,165 @@
+            return Response.status(Response.Status.UNAUTHORIZED).build();
+        } catch (final IndexOutOfBoundsException | IllegalArgumentException e) {
+            this.logger.error("The requested offset is out of bounds: {}", offset);
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        } catch (final Exception e) {
+            this.logger.error("Unexpected exception while generating matches JSON: {}", e.getMessage());
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * Calculates the last index based on the provided {@code offset}, {@code limit}, and size of
+     * {@code matchedCluster}.
+     *
+     * @param matchedCluster the {@link MatchedPatientClusterView} object containing the matches for requested patient
+     * @param offset the offset for the match data to be returned
+     * @param limit the limit for the number of matches to be returned
+     * @return the index of the last match to be returned
+     */
+    private int getLastIndex(@Nonnull final MatchedPatientClusterView matchedCluster, final int offset, final int limit)
+    {
+        final int totalSize = matchedCluster.size();
+        final int lastItemIdx = totalSize - 1;
+        final int requestedLast = limit >= 0 ? offset + limit - 2 : lastItemIdx;
+        return requestedLast <= lastItemIdx ? requestedLast : lastItemIdx;
+    }
+
+    /**
+     * Returns the JSON representation matching patients for {@code reference}. Throws a {@link WebApplicationException}
+     * if {@code reference} is blank.
+     *
+     * @param reference the reference patient ID
+     * @param offset the offset for the returned match results
+     * @param limit the limit for the number of matches to return
+     * @param reqNo the request number
+     * @return a JSON representation of matches for patient with ID {@code reference}
+     * @throws WebApplicationException if {@code reference} is blank, or patient with that ID does not exist
+     */
+    private Response getMatchesResponse(
+        @Nullable final String reference,
+        final int offset,
+        final int limit,
+        final int reqNo)
+    {
+        if (StringUtils.isBlank(reference)) {
+            this.logger.error("No reference patient ID was provided.");
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+        final Patient patient = this.repository.get(reference);
+
+        if (patient == null) {
+            this.logger.error("Patient with ID: {} could not be found.", reference);
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+        final List<PatientSimilarityView> matches = this.similarPatientsFinder.findSimilarPatients(patient);
+        final boolean isManager = ((SecurePatientAccess) this.manager.getPatientAccess(patient)).hasAccessLevel(MANAGE);
```

Comment:
Matches are not saved to the matching notification table here, but they should be. A good fix is to use `LocalMatcher` or even higher-level match-finding code to find matches here, so that all the low-level work implemented there (such as saving matches) is reused